rfd/0096-securing-docker-images.md

```diff
@@ -0,0 +1,378 @@
+results will by uploaded to the GitHub Code Scanning service. From there, our Panther 
+SIEM can observe any alerts and instigate remedial action.
+
+### Software Bill-of-Materials
+
+We can use `syft` to generate a SBOM and `cosign` to attach it to the resulting image. Any
+software components add to the image must be automatically discoverable by `syft` (for 
+example, making sure the package control file is included in `/var/lib/dpkg/status.d` 
+for Debian packages).
+
+### Release Build process summary
+
+This process describes how an image is built during a full Teleport release. 
+
+```mermaid
+graph TD
+    pull_base[Pull distroless<br/>base image]
+    signature_valid?{is cosign<br/>signature valid?}
+    teleport_deb[/Teleport Debian Package<br/>from CI/]
+    third_party_debs[/Third party Debian packages<br/>from Debian package repository/]
+    df[/Dockerfile from<br/>Teleport git repository/]
+    base[/Distroless base image/]
+    candidate_image[/Candidate Telport image/]
+    smoke_test[Smoke test to see if<br/>Teleport starts in container image]
+    smoke_test_pass?{Smoke test<br/>passes?}
+    push_internal[Push Candidate Image<br/>to private ECR]
+    trivy_scan[Scan Candidate Image<br/>with trivy]
+    build[$ docker build ...] 
+    upload_results[Upload scan results<br/>to GitHub Code Scanning]
+
+    fail_build[Fail the build]
+    sign_image[Sign Candidate Image<br/>with Cosign]
+    promote_build[/Release engineer<br/>promotes build/]
+    promote_image[Candidate Image, Signature & SBOM<br/>copied to Public ECR]
+    build_ok?{Does all go well<br/>with the rest of<br/>the build?}
+    sbom[Generate SBOM with syft]
+    attach_sbom[Attach SBOM to<br/>Candidate Image with cosign]
+    
+    pull_base --> signature_valid?
+    signature_valid? -- yes --> base    
+    signature_valid? -- no --> fail_build
+    teleport_deb --> build
+    third_party_debs --> build
+    df --> build
+    base --> build    
+    build --> candidate_image
+
+    candidate_image --> smoke_test   
+    smoke_test --> smoke_test_pass?
+    smoke_test_pass? -- yes --> push_internal
+    smoke_test_pass? -- no --> fail_build
+    push_internal --> trivy_scan
+    push_internal --> sbom
+    trivy_scan --> upload_results 
+    push_internal --> sign_image
+    sbom --> attach_sbom
+    upload_results --> code_scanning_alerts[(GitHub Code<br/>Scanning alerts)]
+    code_scanning_alerts --> panther[(Panther SIEM)]
+    
+    sign_image --> build_ok?
+    upload_results --> build_ok?
+    attach_sbom --> build_ok?
+    
+    promote_build --> promote_image
+    build_ok? -- yes --> promote_build
+    build_ok? -- no --> fail_build
+```    
+
+### Debug Images
+
+Troubleshooting a distroless image is hard, as there are no tools baked into
+the image to aid in debugging a deployment.
+
+In addtion to the main distroless image sets, the `distroless` team also supplies 
+a `debug`-tagged image that includes `busybox` (that is, a basic shell and some
```

Comment:
If we hold the debug images to the same standard as defined above, we will lose some of the benefits of reduced internal efforts (more surface area potentially requiring prompt responses).

I agree distroless is hard to debug and we should consider this.  It may be worth reconsidering `apko` or other options that provide a single, minimized, but debugable distro rather than trying to hold two release streams to the same rigid standard.

Reply:
Right. This is a bit of a worry that we would say that our happy path image is secure, but unhappy path is "use at your own risk". Attacker would always choose the path of least resistance. They would attempt to infiltrate the weak link - the debug container, and patiently wait for the Teleport's crash on someone infrastructure.

That is why I think we should provide a debug container that is as secure as our production. Can we use apko to generate a secure debug container? If that's the case, may be reconsider apko for the main use-case as well?

TLDR: Attackers will pick our weakest link, which is a debug container.

Reply:
I may be missing something but our debug image will also be based on distroless and signed, will just include extra tools like busybox. Apko also generates images with extra stuff in them but it's a some lesser-known 3rd party tool while our current debug image will be based on Google's official distroless debug image which sounds like a better choice? I guess the downside is that we have 2 images but I don't think it's fair to say that the "debug" image (maybe name "debug" is a bit unfortunate) is held to a lesser standard than the "slim" image. @tcsc correct me if I'm wrong.

Reply:
To better articulate the advantages I see in going distroless:
1. Smaller in size (less storage and transfer costs)
2. Reduced footprint results in less dependency CVE's that require team interruptions to address
3. Less tooling for attackers to be able to utilize once an escape has occurred
4. Easier to harden (basically already there)

Using a `debug` image compromises each one of these qualities (to varying degrees):
1. More unused contents that we have to store and transfer
2. More tools, which if we hold `debug` to the same standard could result in more team interruptions to keep it updated and secure
3. Just the availability of a shell is meaningful, but other GNU tooling is also present
4. The act of hardening a `debug` image may render the `debug` capability hindered to the point that it could never adhere to the same standard of security

TL:DR; The act of hardening and keeping `debug` up to date will be more effort than if we have a single image that we focus our efforts on.  I really like `distroless`, but I dislike the idea of eroding it's benefits by needing to maintain a second `debug` image.

That said, there is no technical limitation, or unmanageable security risk here.  I just believe that this solution will result in more effort for the team than if we adopted a single base image solution.

Reply:
@jentfoo @r0mant if our `debug` image will also use distroless, signed and scanned, I have no objections to this plan. But we should not say `use at your own risk` then. We will be holding our debug images to the same standard, which I think is appropriate. 

If that's the case, I have no objections.

Reply:
Taking a step back, our motivation for the debug image is:

[This is what we're offering to customer today](https://github.com/gravitational/teleport/blob/e4c97db936239ca3bac1d8c53956829c7ebb64c9/build.assets/charts/Dockerfile) -- a fat image with useful debugging tools in it.

We want to stop doing that by default, but we can't have an instant cutover.  Therefore, we *must* offer a image compatible with what were putting out today in the short term.  It won't be up to the same hardening standards as the minimial images -- by design.

Reply:
Sounds good to me, particularly if the long term strategy is to have no debug image.  I just was trying to highlight that if we feel a debug version is necessary, we may want to consider just focusing our efforts on a single, secure, but debug-able image.